File: Quorum/Library/Standard/Plugins/javascript/Libraries/Language/Object.js
var globalStaticHash = 1;
const globalSharedClasses = new Map();

function global_Get_Shared_Class(value) {
    return globalSharedClasses.get(value);
}

function global_Add_Shared_Class(key, value) {
    globalSharedClasses.set(key, value);
    return value;
}

function global_Remove_Shared_Class(key) {
    globalSharedClasses.delete(key);
}

function global_Empty_Shared_Classes() {
    globalSharedClasses.clear();
}

function global_cast_text_to_boolean(value) {
	if(value == "true") {
    		return true;
	} else if (value == "false") {
   		return false;
	} else {
		throw "Error: Cast Error: An error occurred while casting";
	}
}

function global_number_converter_(value) {
    var result = value % 1 == 0 ? value.toFixed(1) : value;
    result = result + "";
    result = result.replace("e+", "E");
    result = result.replace("e-", "E-");
    return result;
}

function global_truncate_(value) {
    if(value > 0) {
        return Math.floor(value);
    } else {
        return Math.ceil(value);
    }
}

function global_InstanceOf(variable, name) {
    if (variable === null || variable === undefined)
        return false;
    
    var value = variable.parentNames_;
    if(variable.prototype != null) {
        value = variable.prototype.parentNames_;
    }
    for	(index = 0; index < value.length; index++) {
        if(value[index] == name) {
            return true;
        }
    }
    return false;
}

function global_parseInteger(text) {
    text = text.trim();
    if(text == null) {
        var error = new quorum_Libraries_Language_Errors_CastError_();
        throw error;
    }
    if(text.length == 0) {
        var error = new quorum_Libraries_Language_Errors_CastError_();
        throw error;
    }
    var index = 0;
    var result = 0;
    var isMinus = false;
    if(text.charAt(index) == '+') {
        index = 1;
    } else if(text.charAt(index) == '-') {
        isMinus = true;
        index = 1;
    }
    for (; index < text.length; index++) {
        var char2 = text.charCodeAt(index) - 48;
        if(char2 >= 0 && char2 < 10) {
            result = result * 10;
            result = result + char2;
        } else { /* unlike typical javascript, throw an exception from Quorum */
            var error = new quorum_Libraries_Language_Errors_CastError_();
            throw error;
        }
    }
    if(isMinus) {
        result = result * -1;
    }
    return result;
}

function global_parseNumber(text) {
    text = text.trim();
    if (text == null) {
        var error = new quorum_Libraries_Language_Errors_CastError_();
        throw error;
    }
    if (text.length == 0) {
        var error = new quorum_Libraries_Language_Errors_CastError_();
        throw error;
    }
    if(/^(\-|\+)?([0-9]+(\.[0-9]+)?((e|E)(\-|\+)?[0-9]+)?|Infinity)$/.test(text)) {
      return Number(text);
    } else {
        var error = new quorum_Libraries_Language_Errors_CastError_();
        throw error;
    }
}

function global_CheckCast(from, to) {
    if (from === null || from === undefined)
        return from;
    var names = from.parentNames_;
    for(i = 0; i < names.length; i++) {
        if(names[i] == to) {
            return from;
        }
    }
    var error = new quorum_Libraries_Language_Errors_CastError_();
    throw error;
}

//this function needs work and cross-browser testing to make sure it 
//works as expected everywhere. As of now, exceptions in the JavaScript
//version will have incorrect sometimes.
function global_ErrorCheck(error) {
    names = error.parentNames_;
    if (names == null) {
        if(error instanceof TypeError) {
            return new quorum_Libraries_Language_Errors_UndefinedObjectError_();
        } else {
            return new quorum_Libraries_Language_Errors_Error_();
        }
    } else {
        return error; /*it's already a quorum error*/
    }
}

function global_GetValue_(value, type) {
    if (value == null) {
        if(type == "number" || type == "boolean" || type == "integer") {
            throw "party foul";
        } else {
            return null;
        }
    } else {
        return value.GetValue();
    }
}

File: Quorum/Library/Standard/Plugins/python/Libraries/Language/Object.py
globalSharedClasses = {}

def Global_Get_Shared_Class(key):
	return globalSharedClasses.get(key)

def Global_Add_Shared_Class(key, value):
	return globalSharedClasses.setdefault(key, value)

def Global_Remove_Shared_Class(key):
	globalSharedClasses.pop(key);

def global_Empty_Shared_Classes():
	globalSharedClasses.clear()
	
def Global_Cast_Text_To_Boolean(value):
	if value == "true":
    		return True
	elif value == "false":
		return False
	else:
		raise Exception("Error: Cast Error: An error occurred while casting")
		
def Global_Cast_To_Text(value):
	return str(value)
	
def Global_Cast_Text_To_Integer(value):
	try:
		return int(value)
	except ValueError as e:
		raise quorum_Libraries_Language_Errors_CastError_()
	
def Global_Cast_Text_To_Number(value):
	try:
		return float(value)
	except ValueError as e:
		raise quorum_Libraries_Language_Errors_CastError_()

def Global_Cast_Boolean_To_Text(value):
	if value == True:
		return "true"
	else:
		return "false"

def Global_Cast_Boolean_To_Number(value):
	if value == True:
		return float(1.0)
	else:
		return float(0.0)

def Global_Cast_Boolean_To_Integer(value):
	if value == True:
		return int(1)
	else:
		return int(0)
		
def Global_Cast_Integer_To_Text(value):
	return str(value)
	
def Global_Cast_Number_To_Text(value):
	value = str(value)
	value = value.replace("e+", "E")
	return value
	
def Global_Cast_Integer_To_Number(value):
	return float(value)
	
def Global_Cast_Number_To_Integer(value):
	return int(value)
		
def  Global_GetValue_(value, type):
	if (value == None):
		if(type == "number" or type == "boolean" or type == "integer"):
			raise ValueError("Cannot get a value from the type: " + type)
		else:
			return None
	else:
		return value.GetValue()
		
def Global_CheckCast(from_, to):
	if (from_ is None or from_ == None):
		return from_
	names = from_.inheritance_;
	for element in names:
		if(element == to):
			return from_
    
	raise quorum_Libraries_Language_Errors_CastError_()
	
def Exception_(e):
	val = e
	if isinstance(e, quorum_Libraries_Language_Errors_Error_):
		return e
	elif isinstance(e, AttributeError):
		if "NoneType" in str(e): 
			val = quorum_Libraries_Language_Errors_UndefinedObjectError_()
		else:
			val = quorum_Libraries_Language_Errors_Error_()
	else:
		val = quorum_Libraries_Language_Errors_Error_()
	
	message = str(e)
	val.Set_Libraries_Language_Errors_Error__errorMessage_(message)
	
	return val